tree/ntuple/v7/src/RNTupleSerialize.cxx

```diff
@@ -986,46 +965,234 @@ RResult<std::uint32_t> ROOT::Experimental::Internal::RNTupleSerializer::Deserial
+   if (forHeaderExtension) {
+      if (auto xHeader = desc.GetHeaderExtension()) {
+         nFields = xHeader->GetNFields();
+         nColumns = xHeader->GetNPhysicalColumns();
+         nAliasColumns = xHeader->GetNLogicalColumns() - xHeader->GetNPhysicalColumns();
+         fieldListOffset = context.GetHeaderExtensionOffset();
+      }
+   } else {
+      nFields = desc.GetNFields() - 1;
+      nColumns = desc.GetNPhysicalColumns();
+      nAliasColumns = desc.GetNLogicalColumns() - desc.GetNPhysicalColumns();
+   }
+   const auto &onDiskFields = context.GetOnDiskFieldList();
+   std::span<const DescriptorId_t> fieldList{onDiskFields.begin() + fieldListOffset, onDiskFields.end()};
 
    auto frame = pos;
-   R__ASSERT(desc.GetNFields() > 0); // we must have at least a zero field
-   pos += SerializeListFramePreamble(desc.GetNFields() - 1, *where);
-   pos += SerializeFieldTree(desc, context, *where);
+   pos += SerializeListFramePreamble(nFields, *where);
+   pos += SerializeFieldList(desc, fieldList, /*firstOnDiskId=*/fieldListOffset, context, *where);
    pos += SerializeFramePostscript(buffer ? frame : nullptr, pos - frame);
 
    frame = pos;
-   pos += SerializeListFramePreamble(desc.GetNPhysicalColumns(), *where);
-   pos += SerializeColumnListV1(desc, context, *where);
+   pos += SerializeListFramePreamble(nColumns, *where);
+   pos += SerializeColumnListV1(desc, fieldList, context, *where);
    pos += SerializeFramePostscript(buffer ? frame : nullptr, pos - frame);
 
    frame = pos;
-   auto nAliasColumns = desc.GetNLogicalColumns() - desc.GetNPhysicalColumns();
    pos += SerializeListFramePreamble(nAliasColumns, *where);
-   pos += SerializeAliasColumnList(desc, context, *where);
+   pos += SerializeAliasColumnList(desc, fieldList, context, *where);
    pos += SerializeFramePostscript(buffer ? frame : nullptr, pos - frame);
 
    // We don't use extra type information yet
    frame = pos;
    pos += SerializeListFramePreamble(0, *where);
    pos += SerializeFramePostscript(buffer ? frame : nullptr, pos - frame);
+   return static_cast<std::uint32_t>(pos - base);
+}
+
+RResult<std::uint32_t>
+ROOT::Experimental::Internal::RNTupleSerializer::DeserializeSchemaDescription(const void *buffer, std::uint32_t bufSize,
+                                                                              RNTupleDescriptorBuilder &descBuilder)
+{
+   auto base = reinterpret_cast<const unsigned char *>(buffer);
+   auto bytes = base;
+   auto fnBufSizeLeft = [&]() { return bufSize - (bytes - base); };
+   RResult<std::uint32_t> result{0};
+
+   std::uint32_t frameSize;
+   auto frame = bytes;
+   auto fnFrameSizeLeft = [&]() { return frameSize - (bytes - frame); };
+
+   std::uint32_t nFields;
+   result = DeserializeFrameHeader(bytes, fnBufSizeLeft(), frameSize, nFields);
+   if (!result)
+      return R__FORWARD_ERROR(result);
+   bytes += result.Unwrap();
+   const std::uint32_t fieldIdRangeBegin = descBuilder.GetDescriptor().GetNFields() - 1;
```

Comment:
Add a comment why the descriptor's field list is never empty (the zero field is always added before `DeserializeSchemaDescription` is called)